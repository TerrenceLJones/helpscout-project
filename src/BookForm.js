import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';

export default class BookForm extends Component {
  static propTypes = {
    book: PropTypes.object
  }

  state = {
    category: '',
    id: '',
    image: '',
    imagePreviewUrl: '',
    title: ''
  }

  componentDidMount() {
    const { book } = this.props;

    if(book) {
      this.setState({
        category: book.category,
        id: book.id,
        imagePreviewUrl: book.image,
        title: book.title
      });
    }
  }

  handleTitleChange = ({ target: { value } }) => {
    this.setState({ title: value })
  }

  handleCategoryChange = ({ target: { value } }) => {
    this.setState({ category: value })
  }

  handleImageChange = ({ target: { files } }) => {
    let reader = new FileReader();
    let file = files[0];

    reader.onloadend = () => {
      this.setState({
        image: file,
        imagePreviewUrl: reader.result
      });
    }

    reader.readAsDataURL(file)
  }

  handleSubmit = (e) => {
    e.preventDefault();
    console.log('handleSubmit', this.state);
  }

  handleDeleteBook = (e) => {
    e.preventDefault();
    console.log('handleDeleteBook', this.state);
  }

  getHeader() {
    const headerText = this.state.id ? 'Edit Book' : 'Create a New Book';

    return (
      <header>
        <h1>{ headerText }</h1>
        <Link to={ `/books/${this.state.id}` }>Cancel</Link>
      </header>
    );
  }

  getImagePreview = () => {
    let { imagePreviewUrl } = this.state;

    if (!imagePreviewUrl) {
      return <div>Please select an Image for Preview</div>;
    }

    return <img src={ imagePreviewUrl } alt="Your book&#39;s cover image."/>;
  }

  getSaveButton() {
    if(!this.state.id) {
      return <button type="submit">Create</button>;
    }

    return <button type="submit">Save</button>;
  }

  getDeleteButton() {
    if(!this.state.id) {
      return null;
    }

    return <button type="button" onClick={ this.handleDeleteBook }>Delete</button>;
  }

  render() {
    return (
      <div>
        { this.getHeader() }

        <form onSubmit={ this.handleSubmit }>

          <label htmlFor="title">Title</label>
          <input
            id="title"
            onChange={ this.handleTitleChange }
            type="text"
            value={ this.state.title }
          />

          <label htmlFor="category">Category</label>
          <input type="text"
            id="category"
            onChange={ this.handleCategoryChange }
            value={ this.state.category }
          />

          { this.getImagePreview() }

          <input
            accept="image/*"
            id="image"
            onChange={ this.handleImageChange }
            type="file"
          />

          { this.getSaveButton() }

          { this.getDeleteButton() }
        </form>
      </div>
    );
  }
}
