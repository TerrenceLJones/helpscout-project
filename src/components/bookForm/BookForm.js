import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import styles from 'components/bookForm/BookForm.module.css';

import { getPlaceHolderImageUrl } from 'helpers/placeHolderImage';

const BookForm = class BookForm extends Component {
  static propTypes = {
    book: PropTypes.object,
    onSubmit: PropTypes.func.isRequired,
    ondDelete: PropTypes.func
  }

  state = {
    author: '',
    category: '',
    id: '',
    image: '',
    imagePreviewUrl: '',
    title: ''
  }

  constructor(props) {
    super(props)

    this.imageUploaderRef = React.createRef();
  }

  componentDidMount() {
    const { book } = this.props;

    if(book) {
      this.setState({
        author: book.author,
        category: book.category,
        id: book.id,
        image: book.image,
        imagePreviewUrl: book.image,
        title: book.title
      });
    }
  }

  handleAuthorChange = ({ target: { value } }) => {
    this.setState({ author: value })
  }

  handleTitleChange = ({ target: { value } }) => {
    this.setState({ title: value })
  }

  handleCategoryChange = ({ target: { value } }) => {
    this.setState({ category: value })
  }

  handleAddBookCover = () => {
    const node = this.imageUploaderRef.current;

    if(node) {
      node.click();
    }
  }

  handleImageChange = ({ target: { files } }) => {
    let reader = new FileReader();
    let file = files[0];

    reader.onloadend = () => {
      this.setState({
        image: reader.result,
        imagePreviewUrl: reader.result
      });
    }

    reader.readAsDataURL(file)
  }

  handleSubmit = (e) => {
    e.preventDefault();

    this.props.onSubmit(this.state);
  }

  getImagePreview = () => {
    const { imagePreviewUrl } = this.state;
    const title = this.props.book && this.props.book.title;
    const backgroundImage = imagePreviewUrl || getPlaceHolderImageUrl(title);
    const containerClasses = classNames(styles.container, 'card shadow-sm');
    const imagePreviewClasses = classNames(styles.linkButton, styles.image);
    const imagePreviewButtonClasses = classNames(styles.linkButton, styles.buttonText);

    return(
      <div className={ containerClasses }>
        <div className={ styles.imageContainer }>
          <button
            style={ { backgroundImage: `url("${backgroundImage}")` } }
            className={ imagePreviewClasses }
            type="button"
            onClick={ this.handleAddBookCover }
          />
        </div>

        <div className="card-body">
          <input
            accept="image/*"
            className={ styles.bookImageAddButton }
            id="image"
            onChange={ this.handleImageChange }
            ref={ this.imageUploaderRef }
            style={ { display: 'none' } }
            type="file"
          />

          <button
            className={ imagePreviewButtonClasses }
            type="button"
            onClick={ this.handleAddBookCover }
          >
            Add Book Cover
          </button>
        </div>
      </div>
    );
  }

  getSaveButton() {
    const text = this.state.id ? 'Save' : 'Create'

    return <button className="btn btn-success btn-block" type="submit">{ text }</button>;
  }

  getDeleteButton() {
    if(!this.state.id) {
      return null;
    }

    return (
      <button
        className="btn btn-danger btn-block"
        type="button"
        onClick={ () => this.props.onDelete(this.state.id) }>
        Delete
      </button>);
  }

  render() {
    return (
      <div>
        <form onSubmit={ this.handleSubmit }>

          <div className="row mb-4">
            <div className="col-md-6 mb-4">
              { this.getImagePreview() }
            </div>

            <div className="col-md-6">
              <div className="form-group">
                <label htmlFor="title">Title</label>
                <input
                  className="form-control"
                  id="title"
                  onChange={ this.handleTitleChange }
                  type="text"
                  value={ this.state.title }
                />
              </div>

              <div className="form-group">
                <label htmlFor="title">Author</label>
                <input
                  className="form-control"
                  id="title"
                  onChange={ this.handleAuthorChange }
                  type="text"
                  value={ this.state.author }
                />
              </div>

              <div className="form-group">
                <label htmlFor="category">Category</label>
                <input type="text"
                  className="form-control"
                  id="category"
                  onChange={ this.handleCategoryChange }
                  value={ this.state.category }
                />
              </div>

              { this.getSaveButton() }
              { this.getDeleteButton() }

            </div>
          </div>
        </form>
      </div>
    );
  }
}

export {
  BookForm
}

export default BookForm;
