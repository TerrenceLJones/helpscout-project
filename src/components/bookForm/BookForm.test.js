import React from 'react';
import { shallow } from 'enzyme';

import { BookForm } from './BookForm';


it('renders without crashing', () => {
    expect(shallow(<BookForm
        book={ {} }
        onSubmit={ jest.fn() }
        onDelete={ jest.fn() }
    />)).toMatchSnapshot();
});

it('calls handleSubmit after triggering submit', () => {});
it('calls handleDelete after triggering delete', () => {});
