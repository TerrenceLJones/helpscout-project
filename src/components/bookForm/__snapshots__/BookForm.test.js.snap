// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <BookForm
    book={Object {}}
    handleDelete={[MockFunction]}
    handleSubmit={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <form
        onSubmit={[Function]}
      >
        <label
          htmlFor="title"
        >
          Title
        </label>
        <input
          id="title"
          onChange={[Function]}
          type="text"
        />
        <label
          htmlFor="category"
        >
          Category
        </label>
        <input
          id="category"
          onChange={[Function]}
          type="text"
        />
        <div>
          Please select an Image for Preview
        </div>
        <input
          accept="image/*"
          id="image"
          onChange={[Function]}
          type="file"
        />
        <button
          type="submit"
        >
          Create
        </button>
      </form>,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <label
            htmlFor="title"
          >
            Title
          </label>,
          <input
            id="title"
            onChange={[Function]}
            type="text"
          />,
          <label
            htmlFor="category"
          >
            Category
          </label>,
          <input
            id="category"
            onChange={[Function]}
            type="text"
          />,
          <div>
            Please select an Image for Preview
          </div>,
          <input
            accept="image/*"
            id="image"
            onChange={[Function]}
            type="file"
          />,
          <button
            type="submit"
          >
            Create
          </button>,
          null,
        ],
        "onSubmit": [Function],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Title",
            "htmlFor": "title",
          },
          "ref": null,
          "rendered": "Title",
          "type": "label",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "id": "title",
            "onChange": [Function],
            "type": "text",
            "value": undefined,
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Category",
            "htmlFor": "category",
          },
          "ref": null,
          "rendered": "Category",
          "type": "label",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "id": "category",
            "onChange": [Function],
            "type": "text",
            "value": undefined,
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Please select an Image for Preview",
          },
          "ref": null,
          "rendered": "Please select an Image for Preview",
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "accept": "image/*",
            "id": "image",
            "onChange": [Function],
            "type": "file",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Create",
            "type": "submit",
          },
          "ref": null,
          "rendered": "Create",
          "type": "button",
        },
        null,
      ],
      "type": "form",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <form
          onSubmit={[Function]}
        >
          <label
            htmlFor="title"
          >
            Title
          </label>
          <input
            id="title"
            onChange={[Function]}
            type="text"
          />
          <label
            htmlFor="category"
          >
            Category
          </label>
          <input
            id="category"
            onChange={[Function]}
            type="text"
          />
          <div>
            Please select an Image for Preview
          </div>
          <input
            accept="image/*"
            id="image"
            onChange={[Function]}
            type="file"
          />
          <button
            type="submit"
          >
            Create
          </button>
        </form>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <label
              htmlFor="title"
            >
              Title
            </label>,
            <input
              id="title"
              onChange={[Function]}
              type="text"
            />,
            <label
              htmlFor="category"
            >
              Category
            </label>,
            <input
              id="category"
              onChange={[Function]}
              type="text"
            />,
            <div>
              Please select an Image for Preview
            </div>,
            <input
              accept="image/*"
              id="image"
              onChange={[Function]}
              type="file"
            />,
            <button
              type="submit"
            >
              Create
            </button>,
            null,
          ],
          "onSubmit": [Function],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Title",
              "htmlFor": "title",
            },
            "ref": null,
            "rendered": "Title",
            "type": "label",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "id": "title",
              "onChange": [Function],
              "type": "text",
              "value": undefined,
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Category",
              "htmlFor": "category",
            },
            "ref": null,
            "rendered": "Category",
            "type": "label",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "id": "category",
              "onChange": [Function],
              "type": "text",
              "value": undefined,
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Please select an Image for Preview",
            },
            "ref": null,
            "rendered": "Please select an Image for Preview",
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "accept": "image/*",
              "id": "image",
              "onChange": [Function],
              "type": "file",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Create",
              "type": "submit",
            },
            "ref": null,
            "rendered": "Create",
            "type": "button",
          },
          null,
        ],
        "type": "form",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
